File: Aed_2/Trabalhos praticos/TP_01/Q08/Soma.c
#include<stdio.h>
#include<stdlib.h>
#include<stdbool.h>
#include<string.h>
#include<ctype.h>

/**
 * Método recursivo para somar termos de uma String
 * @param num String de entrada padrão
 * @param i Índice de controle
 * @param sum Valor da soma
 * @return Valor da soma
 */
int SomaTermos(char* num, int i, int sum){

    // Definir dados
    int len = strlen(num);

    if(i == len){

        return sum;
    } else {

        sum += (int) num[i] - '0';

        return SomaTermos(num, i + 1, sum);
    }
}

/**
 * Método principal que lê strings e soma os seus termos.
 * O programa encerra quando a entrada for "FIM".
 * @param args Argumentos da linha de comando.
 */
int main(){

    // Definir dados
    int num;
    char str[256];

    fgets(str, 256, stdin);
    str[strcspn(str, "\n")] = '\0';

    while(strcmp(str, "FIM") != 0){

        printf("%d\n", SomaTermos(str, 0, 0));

        fgets(str, 256, stdin);
        str[strcspn(str, "\n")] = '\0';
    }
    

    return 0;
}

File: Aed_2/Trabalhos praticos/TP_01/Q08/Soma.java


import java.util.Scanner;

/**
 * Classe para somar termos de uma String.
 */
public class Soma {
    
    /**
     * Método recursivo encapsulado para somar termos de uma String
     * @param num String de entrada padrão
     * @return Valor da soma
     */
    public static int SomaTermos(String num){

        return SomaTermos(num, 0, 0);
    }

    /**
     * Método recursivo para somar termos de uma String
     * @param num String de entrada padrão
     * @param i Índice de controle
     * @param sum Valor da soma
     * @return Valor da soma
     */
    public static int SomaTermos(String num, int i, int sum){
        
        // Passo base
        if(i == num.length()){

            return sum;
        } else {

            char c = num.charAt(i);
            if(c >= '0' && c <= '9'){

                sum += (int) c - '0';
            }
            
            // Passo recursivo
            return SomaTermos(num, i + 1, sum);
        }
        
    }

    /**
     * Método principal que lê strings e soma os seus termos.
     * O programa encerra quando a entrada for "FIM".
     * @param args Argumentos da linha de comando.
     */
    public static void main(String[] args) {
        
        try ( 
            
            // Definir dados
            Scanner scanner = new Scanner(System.in)) {
            String n;
            
            // Leitura de dados
            n = scanner.nextLine();
            
            while(!n.equals("FIM")){
                
                // Mostrar dados
                System.out.println(SomaTermos(n));
                
                n = scanner.nextLine();
            }
        }
    }
}
